homeassistant/components/toon.py

```diff
@@ -0,0 +1,151 @@
+        """Initialize toon."""
+        from toonlib import Toon
+
+        # Creating the class
+
+        toon = Toon(username, password)
+
+        self.toon = toon
+        self.gas = gas
+        self.solar = solar
+        self.data = {}
+
+        self.last_update = datetime.min
+        self.update()
+
+    def update(self):
+        """Update toon data."""
+        now = datetime.now()
+        if now - self.last_update < timedelta(seconds=5):
+            return
+
+        self.last_update = now
+
+        self.data['power_current'] = self.toon.power.value
+        self.data['power_today'] = round(
+            (float(self.toon.power.daily_usage) +
+             float(self.toon.power.daily_usage_low)) / 1000, 2)
+        self.data['temp'] = self.toon.temperature
+
+        if self.toon.thermostat_state:
+            self.data['state'] = self.toon.thermostat_state.name
+        else:
+            self.data['state'] = 'Manual'
+
+        self.data['setpoint'] = float(
+            self.toon.thermostat_info.current_set_point) / 100
+        self.data['gas_current'] = self.toon.gas.value
+        self.data['gas_today'] = round(float(self.toon.gas.daily_usage) /
+                                       1000, 2)
+
+        for plug in self.toon.smartplugs:
+            self.data[plug.name] = {'current_power': plug.current_usage,
+                                    'today_energy': round(
+                                        float(plug.daily_usage) / 1000, 2),
+                                    'current_state': plug.current_state,
+                                    'is_connected': plug.is_connected}
+
+        self.data['solar_maximum'] = self.toon.solar.maximum
+        self.data['solar_produced'] = self.toon.solar.produced
+        self.data['solar_value'] = self.toon.solar.value
+        self.data['solar_average_produced'] = self.toon.solar.average_produced
+        self.data['solar_meter_reading_low_produced'] = \
+            self.toon.solar.meter_reading_low_produced
+        self.data['solar_meter_reading_produced'] = \
+            self.toon.solar.meter_reading_produced
+        self.data['solar_daily_cost_produced'] = \
+            self.toon.solar.daily_cost_produced
+
+        for detector in self.toon.smokedetectors:
+            value = '{}_smoke_detector'.format(detector.name)
+            self.data[value] = {'smoke_detector': detector.battery_level,
+                                'device_type': detector.device_type,
+                                'is_connected': detector.is_connected,
+                                'last_connected_change':
+                                detector.last_connected_change}
+
+    def set_state(self, state):
+        """Push a new state to the Toon unit."""
+        self.toon.thermostat_state = state
+        self.update()
+
+    def set_temp(self, temp):
+        """Push a new temperature to the Toon unit."""
+        self.toon.thermostat = temp
+        self.update()
```

Comment:
These are not necessary. The climate entity is set to be polled, so Home Assistant will call the update method after it triggers a service.